mountpoint-s3-crt/src/s3/endpoint_resolver.rs

```diff
@@ -0,0 +1,329 @@
+        Ok(Self { inner })
+    }
+
+    /// Resolve the endpoint with given [RequestContext] and ruleset.
+    pub fn resolve(&self, context: RequestContext) -> Result<ResolvedEndpoint, Error> {
+        // SAFETY: `aws_endpoints_rule_engine_resolve` ensures that it returns a `aws_endpoints_resolved_endpoint`
+        // (which is checked for being non-null later) or it will return an error. So, the out_endpoint is valid and non null.
+        // SAFETY: Next, we handle all the arms `aws_endpoints_resolved_endpoint_get_type` separately.
+        unsafe {
+            let mut out_endpoint: *mut aws_endpoints_resolved_endpoint = ptr::null_mut();
+            aws_endpoints_rule_engine_resolve(self.inner.as_ptr(), context.inner.as_ptr(), &mut out_endpoint)
+                .ok_or_last_error()?;
+            match aws_endpoints_resolved_endpoint_get_type(out_endpoint) {
+                aws_endpoints_resolved_endpoint_type::AWS_ENDPOINTS_RESOLVED_ERROR => {
+                    let mut out_error: aws_byte_cursor = Default::default();
+                    let err = Error::from(aws_endpoints_resolved_endpoint_get_error(out_endpoint, &mut out_error));
+                    Err(err)
+                }
+                aws_endpoints_resolved_endpoint_type::AWS_ENDPOINTS_RESOLVED_ENDPOINT => {
+                    let out_endpoint = out_endpoint.ok_or_last_error()?;
+                    Ok(ResolvedEndpoint { inner: out_endpoint })
+                }
+                _ => unreachable!("Invalid resolved endpoint type"),
+            }
+        }
+    }
+}
+
+impl Drop for RuleEngine {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` is a valid `aws_endpoints_rule_engine`, and on Drop it's safe to decrement
+        // the reference count since we won't use it again through `self`.
+        unsafe {
+            aws_endpoints_rule_engine_release(self.inner.as_ptr());
+        }
+    }
+}
+
+/// Stores context about the endpoint configuration for use by the endpoint resolver
+#[derive(Debug)]
+pub struct RequestContext {
+    inner: NonNull<aws_endpoints_request_context>,
+}
+
+impl RequestContext {
+    /// Creates a new endpoint [RequestContext].
+    pub fn new(allocator: &Allocator) -> Result<Self, Error> {
+        s3_library_init(allocator);
+        // SAFETY: `allocator.inner` is a valid aws_allocator and and we check the return is non-null.
+        // SAFETY: aws_endpoints_request_context_new will acquire a reference to keep it alive after this function call, so it's safe to return an owned version to it.
+        let inner = unsafe { aws_endpoints_request_context_new(allocator.inner.as_ptr()).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Add the parameter to [RequestContext] whose value is in form of string
+    pub fn add_string(&mut self, allocator: &Allocator, name: &OsStr, value: &OsStr) -> Result<(), Error> {
```

Comment:
nit: this is slightly easier to use if it's `name: impl AsRef<OsStr>, value: impl AsRef<OsStr>`. Inside the method, use `name.as_ref()` instead of just `name`. Same thing for `name` in `add_boolean`. You can see examples in methods like `Header::has_header`.

This way you don't have to do the `OsStr::new(...)` dance when you try to call it.